proposals/NNNN-explicit-layout-struct.md

```diff
@@ -0,0 +1,203 @@
+```
+
+Simple usage of packoffset:
+
+```llvm
+; cbuffer packoffset1 {
+;   float x : packoffset(c1.x);
+;   float y : packoffset(c2.y);
+; };
+target("dx.Layout", { i32, i32 }, 40, 16, 36)
+```
+
+packoffset that reorders fields:
+> note: This does not currently work in DXC targeting SPIR-V.
+
+```llvm
+; cbuffer packoffset1 {
+;   float x : packoffset(c2.y);
+;   float y : packoffset(c1.x);
+; };
+target("dx.Layout", { i32, i32 }, 40, 36, 16)
+```
+
+Use of `[[vk::offset()]]`:
+
+```llvm
+; struct vkoffset1 {
+;   float2 a;
+;   [[vk::offset(8)]] float2 b;
+; }
+%__hlsl_vkoffset1 = { <2 x float>, <2 x float> }
+target("spirv.Layout", %__hlsl_vkoffset1, 12, 0, 8)
+
+; struct complex {
+; float r;
+; float i;
+; };
+; struct vkoffset2 {
+;   float2 a;
+;   [[vk::offset(8) complex b;
+; }
+%__hlsl_vkoffset2 = { <2 x float>, { float, float } }
+target("spirv.Layout", %__hlsl_vkoffset1, 16, 0, 8)
+```
+
+## Open questions
+
+#### Arrays in CBuffers
+
+Should we also add a `target("dx.CBufArray", <type>, <size>)` type, rather than
+having the CBuffer logic need to be aware of special array element spacing
+rules?
+
+#### Decaying to non-target types
+
+Operations like `resource.getpointer` can expose us to the contained type of a
+resource, but should that give us an object of the underlying struct or the
+target type? For the former, we would lose the layout, which is problematic.
+For the latter, we need to talk about GEPs.
+
+If we can have pointers of the target type, we'd either need to teach GEP to
+handle these, which is a fairly wide-reaching change, or we would need a set of
+GEP-like intrinsics.
+
+This will need to be resolved in order to use these in StructuredBuffer and for
+most real use-cases of `vk::offset`.
+
+#### Copying in and out of the layout
+
+How do we convert from the Layout types to types without the offsets? We'll
+probably need an intrinsic that does a logical copy from the target type to a
+compatible structure. See https://godbolt.org/z/rh5dvd3E7 for an example.
+
+#### Struct layouts in Clang
+
+Clang uses [ASTContext::getRecordLayout] to get the offset of a member when
+needed. This proposal means that we will not update this class to be aware of
+`vk::offset` and `packedoffset`. The goal of this proposal is to avoid adding
+explicit padding members in the llvm:StructType as is done when handling
+`alignof` in C++.
+
+[ASTContext::getRecordLayout]: https://github.com/llvm/llvm-project/blob/aa9e519b2423/clang/lib/AST/RecordLayoutBuilder.cpp#L3321
+
+If we update `ASTContext::getRecordLayout`, then the explicit padding will be
+added to the llvm::StructType. If we do not update it, then we will have to
+consider every spot in Clang that calls it to see if we need to special case
+the HLSL layout.
+
+Currently known issue are: `offsetof` and `sizeof`. There is a lot more that
+needs to be looked into in Clang. `ASTRecordLayout` is passed around to a lot
+of functions, without the original type.
+
+## Acknowledgments
+
```

Comment:
Another open issue. How will member functions work? This is related to how the `this` pointer will be handled in member functions.